File: web/src/runStage.ts
// The five states a Run passes through, on this side of the wire.
//
// The other adapter is `crates/shared/src/run_stage.rs`; the seam between them
// is the `stage` field's spelling in a Run Log line, which is a contract and
// does not change. Everything a screen needs to know about a stage is decided
// here: what it is called in Chinese, whether the run can still be stopped, and
// why not when it cannot.
//
// Before this module the answers were scattered: `history.ts` interpolated the
// raw wire spelling into a sentence, `run.ts` narrowed three of the five and
// silently dropped the rest, `DesignSystem.tsx` kept its own label table, and
// the stop rule existed only on the server, where the screen learned it by
// clicking and reading a 409.

/** The wire spelling. Identical to `RunStage::as_str` on the Rust side. */
export type RunStage =
  | "PREPARING"
  | "STREAMING"
  | "COMMITTING"
  | "SUCCEEDED"
  | "FAILED";

/**
 * The three the run is *doing something* in, in the order it walks them.
 *
 * The phase line draws these and only these, ending in 「终态待定」: the two
 * terminal stages are not another step along the same track, they are the track
 * finishing, and the run's outcome says which way.
 */
export type RunPhase = "PREPARING" | "STREAMING" | "COMMITTING";

export const RUN_PHASES: readonly RunPhase[] = [
  "PREPARING",
  "STREAMING",
  "COMMITTING",
];

const LABELS: Readonly<Record<RunStage, string>> = {
  PREPARING: "准备中",
  STREAMING: "传输中",
  COMMITTING: "提交中",
  SUCCEEDED: "已完成",
  FAILED: "已失败",
};

function known(raw: string | null): RunStage | null {
  return raw !== null && raw in LABELS ? (raw as RunStage) : null;
}

/** The three-stage subset, for the phase line. `null` for everything else. */
export function runPhase(raw: string | null): RunPhase | null {
  return RUN_PHASES.find((phase) => phase === raw) ?? null;
}

/**
 * What a person should see for this stage, or `null` when there is no stage.
 *
 * **A spelling we do not recognise is shown as it arrived, not swallowed** —
 * the same rule `failureKindLabel` follows, for the same reason: an unknown
 * value means the two ends are on different versions, and that is exactly the
 * moment you want it on screen rather than smoothed away. Nothing is decided
 * off this value, so showing it raw costs nothing but a strange-looking word.
 */
export function stageLabel(raw: RunStage): string;
export function stageLabel(raw: string | null): string | null;
export function stageLabel(raw: string | null): string | null {
  if (raw === null) {
    return null;
  }
  return known(raw) === null ? raw : LABELS[raw as RunStage];
}

/**
 * Whether `source` can still stop this run — the same rule as
 * `RunStage::abort_allowed`, evaluated a second time so the screen can grey the
 * button *before* the round trip instead of after a 409.
 *
 * `null` (the run is accepted but the child has not reported a stage yet) and
 * an unrecognised spelling both answer **no**. Both mean "I do not know what
 * this run is doing", and the server refuses on both, so a lit button would be
 * a promise the next click breaks.
 */
export function abortAllowed(raw: string | null): boolean {
  const stage = known(raw);
  return stage === "PREPARING" || stage === "STREAMING";
}

/**
 * Why the stop button is greyed, in one clause, or `null` when it is not.
 *
 * Three different reasons, deliberately not collapsed into one: at
 * `COMMITTING` the **permission** is gone, at a terminal stage the **process**
 * is gone, and before the first stage there is nothing to stop yet. Only the
 * first is a rule worth explaining; the other two are just facts.
 */
export function abortRefusal(raw: string | null): string | null {
  if (abortAllowed(raw)) {
    return null;
  }
  switch (known(raw)) {
    case "COMMITTING":
      return "已过封口点：暂存表的处置权已经交给目标端";
    case "SUCCEEDED":
    case "FAILED":
      return "运行已经结束，没有可停的进程";
    default:
      return "运行还没拉起来，暂时停不了";
  }
}

File: crates/shared/src/run_stage.rs
//! The five states a Run passes through, and the two rules that hang off them.
//!
//! The stage name crosses a process line: the child writes it into a
//! `stage_changed` Run Log line, and the long-running parent reads it back to
//! decide whether an abort is still allowed. Before this module the two ends met
//! as bare strings — one side spelled them from an enum, the other compared them
//! against literals — so a rename on either side failed silently, in the one
//! place where silence is most expensive.
//!
//! The wire spelling is a contract (`CONTEXT.md` under **Run Log**: the field set
//! is stable, and the acceptance rigs `jq`-match on `.stage == "COMMITTING"`).
//! `as_str` below is that contract, and it is the only place the five words are
//! written down.

use std::fmt;

use serde::{Serialize, Serializer};

/// What the run process is doing right now.
///
/// Named after the work, not after a verdict: `SUCCEEDED` and `FAILED` are the
/// stages a finished process sits in, and they deliberately share their spelling
/// with a run history row's `outcome`, which is the *verdict* over the same two
/// words. The two are separate vocabularies that happen to agree here; nothing
/// converts between them.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RunStage {
    Preparing,
    Streaming,
    Committing,
    Succeeded,
    Failed,
}

impl RunStage {
    /// The closed set, in the order a run walks it. Pinned by a test, like
    /// [`crate::LogEvent::ALL`] — a vocabulary that crosses a process line is
    /// only worth anything if adding to it is a deliberate act.
    pub const ALL: [Self; 5] = [
        Self::Preparing,
        Self::Streaming,
        Self::Committing,
        Self::Succeeded,
        Self::Failed,
    ];

    /// The wire spelling. **Never change one of these** — see the module note.
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Preparing => "PREPARING",
            Self::Streaming => "STREAMING",
            Self::Committing => "COMMITTING",
            Self::Succeeded => "SUCCEEDED",
            Self::Failed => "FAILED",
        }
    }

    /// The direction that used to be missing, and whose absence is why the
    /// parent process compared strings.
    ///
    /// `None` for anything else, including an empty string: an unrecognised
    /// spelling means the two ends are on different versions, and the callers
    /// that make decisions off a stage must all treat that as "I do not know
    /// what this run is doing" rather than guessing. What to *display* for such
    /// a value is a separate question, answered on the front end, which keeps
    /// the raw text so a half-finished upgrade stays visible.
    pub fn parse(text: &str) -> Option<Self> {
        Self::ALL.into_iter().find(|stage| stage.as_str() == text)
    }

    /// Whether `source` may still tell `sink` to discard the staging table.
    ///
    /// This is `CONTEXT.md`'s **Abort** invariant, and this is now its one
    /// implementation: *"It is only ever sent before commit: once `COMMITTING`
    /// is entered, the staging table's disposition has passed wholly to `sink`
    /// and source permanently forfeits the right to abort."*
    ///
    /// The two terminal stages are false for a different reason — there is no
    /// longer a process to stop — which is why refusal wording is chosen by
    /// matching the variant rather than by negating this.
    pub const fn abort_allowed(self) -> bool {
        match self {
            Self::Preparing | Self::Streaming => true,
            Self::Committing | Self::Succeeded | Self::Failed => false,
        }
    }
}

impl fmt::Display for RunStage {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(self.as_str())
    }
}

impl Serialize for RunStage {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_str(self.as_str())
    }
}
